Core/Loot: implement Loot Item Storage

xinef1 opened the pull request:
**Target branch(es):** 3.3.5

- [x] 3.3.5

**Tests performed:** Compiles, simple ingame tests

Created item loot storage, no more synchronous db selects, also fixed buyback case, where stored loot was not removed from db. Added Primary key, and changed field types to be unsigned for table item_loot_money.

Commit:
Created Item Loot Storage, no more synchronous DB selects
Fixed buyback case, where stored loot was not removed from db
Added Primary key, and changed field types to be unsigned for table item_loot_money

Commit:
(c) note

Shauren commented on src/server/game/Loot/LootItemStorage.h:
```diff
@@ -0,0 +1,99 @@
+        ~LootItemStorage();
+
+        static LootItemStorage* instance();
+        static boost::shared_mutex* GetLock();
```
this is a pointer return type, please forward declare namespace boost { class shared_mutex; } - boost headers suck to be included in other headers

Commit:
Forward declare boost::shared_mutex

Commit:
Reduced sections covered by mutex

Commit:
Header fixes

Killyana commented:
This is related to https://github.com/TrinityCore/TrinityCore/issues/8952  ?

xinef1 commented:
No, ill post a fix for that when I'm back. its short one

ariel- committed:
Update:

- Fixed passing SQLTransactions as reference
- Killed vector container (not good for deletions in middle)
- Simplification of constructors and parameters, use LootItem as a proxy struct
- Use RAII for mutexes

ariel- committed:
Updated sql and base

ariel- commented:
I didn't update characters.updates table on purpose, as it depends on the sql name at the moment of the merge.

ariel- commented on src/server/game/Loot/LootItemStorage.cpp:
```diff
@@ -0,0 +1,364 @@
+
+void StoredLootContainer::AddLootItem(LootItem const& lootItem, SQLTransaction& trans)
+{
+    _lootItems.emplace(lootItem.itemid, lootItem);
```
Travis doesn't like this. Any ideas? it compiles just fine on a VM with Clang 3.9 under Ubuntu Server 16.10

Shauren replied:
`_lootItems.emplace(std::piecewise_construct, std::forward_as_tuple(lootItem.itemid), std::forward_as_tuple(lootItem));`

ariel- committed:
Feedback change

ariel- committed:
Initialize StoredLootContainer with container lowguid directly

ariel- committed:
Sql update before merge

Rochet2 commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -13396,7 +13399,12 @@ void Player::RemoveItemFromBuyBackSlot(uint32 slot, bool del)
             if (del)
+            {
                 pItem->SetState(ITEM_REMOVED, this);
+                if (ItemTemplate const* itemTemplate = pItem->GetTemplate())
```
This line has a crash since using `SetState` above will delete the `pItem`
https://github.com/TrinityCore/TrinityCore/blob/9dc3de10f0044c35a95e9b72e7b874f4b5b8e867/src/server/game/Entities/Item/Item.cpp#L674

Must move the SetState below the rest of the IF statements.

ariel- commented:
Gotta love reviews after the PR was opened for almost 3 months
